buildlogger/client.go

```diff
@@ -0,0 +1,319 @@
+		}
+	}
+
+	if opts.Local == nil {
+		opts.Local = send.MakeNative()
+	}
+
+	if opts.MaxBufferSize == 0 {
+		// TODO: figure out ideal default size
+		opts.MaxBufferSize = 4096
+	}
+
+	return nil
+}
+
+// SetExitCode sets the exit code variable.
+func (opts *LoggerOptions) SetExitCode(i int32) { opts.exitCode = i }
+
+// GetLogID returns the unique buildlogger log ID set after NewLogger is
+// called.
+func (opts *LoggerOptions) GetLogID() string {
+	return opts.logID
+}
+
+// NewLogger returns a grip Sender backed by Cedar Buildlogger with level
+// information set.
+func NewLogger(ctx context.Context, name string, l send.LevelInfo, opts *LoggerOptions) (send.Sender, error) {
+	b, err := MakeLogger(ctx, name, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem making new logger")
+	}
+
+	if err := b.SetLevel(l); err != nil {
+		return nil, errors.Wrap(err, "problem setting grip level")
+	}
+
+	return b, nil
+}
+
+// MakeLogger returns a grip Sender backed by Cedar Buildlogger.
+func MakeLogger(ctx context.Context, name string, opts *LoggerOptions) (send.Sender, error) {
+	ts := time.Now()
+
+	if err := opts.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid cedar buildlogger options")
+	}
+
+	var conn *grpc.ClientConn
+	var err error
+	if opts.ClientConn == nil {
+		rpcOpts := []grpc.DialOption{
+			grpc.WithUnaryInterceptor(aviation.MakeRetryUnaryClientInterceptor(10)),
+			grpc.WithStreamInterceptor(aviation.MakeRetryStreamClientInterceptor(10)),
+		}
+		if opts.Insecure {
+			rpcOpts = append(rpcOpts, grpc.WithInsecure())
+		} else {
+			var tlsConf *tls.Config
+			tlsConf, err = aviation.GetClientTLSConfig(opts.CAFile, opts.CertFile, opts.KeyFile)
+			if err != nil {
+				return nil, errors.Wrap(err, "problem getting client TLS config")
+			}
+
+			rpcOpts = append(rpcOpts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConf)))
+		}
+
+		conn, err := grpc.DialContext(ctx, opts.RPCAddress, rpcOpts...)
+		if err != nil {
+			return nil, errors.Wrap(err, "problem dialing rpc server")
+		}
+		opts.ClientConn = conn
+	}
+
+	b := &buildlogger{
+		ctx:    ctx,
+		opts:   opts,
+		conn:   conn,
+		client: internal.NewBuildloggerClient(opts.ClientConn),
+		buffer: []*internal.LogLine{},
+		Base:   send.NewBase(name),
+	}
+
+	if err := b.SetErrorHandler(send.ErrorHandlerFromSender(b.opts.Local)); err != nil {
+		return nil, errors.Wrap(err, "problem setting default error handler")
+	}
+
+	if err := b.createNewLog(ts); err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// Send sends the given message with a timestamp created when the function is
+// called to the Cedar Buildlogger backend. This function buffers the messages
+// created at timestamps until the maximum allowed buffer size is reached at
+// which point a gRPC call is made to send the data to Cedar.
+func (b *buildlogger) Send(m message.Composer) {
+	ts := time.Now()
```

Comment:
Do we need to add any handling of a thread attempting to send while a different one is closing the logger?

Reply:
Added a check below, once Close has been called subsequent calls to Send will fail. Also Close now has grabs the mutex.